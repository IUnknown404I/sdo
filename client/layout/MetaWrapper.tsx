import { useRouter } from 'next/router';
import React, { ReactElement } from 'react';
import { useTypedSelector } from '../redux/hooks';
import logapp from '../utils/logapp';
import { checkProductionMode } from '../utils/utilityFunctions';

/**
 * @IUnknown404I App Wrapper for sending meta data to the server.
 * @param param children as inner components of the wrapper.
 * @returns App Wrapper with isolated logic (per-pages or router-object change logic).
 */
const MetaWrapper = ({ children }: { children: ReactElement | ReactElement[] }): ReactElement => {
	const router = useRouter();
	const auth = useTypedSelector(state => state.auth.state);
	const sessionState = useTypedSelector(state => state.auth.sessionState);
	const axiosInstance = useTypedSelector(state => state.axiosInstance?.instance);
	const accessTokenDTO = useTypedSelector(state => state.accessToken);

	const sendMetaInformationToServer = () => {
		if (!checkProductionMode() || !auth || accessTokenDTO?.access_token == null) return;
		try {
			logapp.log('[#MetaInfoWrapper]: sending metadata', router.asPath);
			axiosInstance
				.put('users/leave-meta', { lastPage: router.asPath })
				.then(res => logapp.log('[# MetaInfoWrapper]: successfully sent data, got response = ', res))
				.catch(e => logapp.log('[# MetaInfoWrapper]: cant send info, got the err: ', e));
		} catch (e) {}
	};

	React.useEffect(() => {
		if (
			auth &&
			sessionState === 'active' &&
			accessTokenDTO.access_token != null &&
			!!router?.asPath &&
			!router?.asPath?.includes('/[')
		)
			sendMetaInformationToServer();
	}, [router, router.asPath]);

	return <>{children}</>;
};

export default MetaWrapper;
