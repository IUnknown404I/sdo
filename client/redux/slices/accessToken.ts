import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import logapp from '../../utils/logapp';
import { LoginFullObjI } from '../api';
import { RootState } from '../store';
import AuthThunks from '../thunks/auth';

export interface AccessTokenI {
	access_token?: string;
	expires_in?: string;
}

export const initialAccessTokenState: AccessTokenI = {
	access_token: undefined,
	expires_in: undefined,
};

export const accessTokenSlice = createSlice({
	name: 'accessToken',
	initialState: initialAccessTokenState,
	reducers: {
		*
	},
	extraReducers: builder => {
		builder.addCase(AuthThunks.fetchAndUpdateTokens.fulfilled, (state, action) => {
			*
		});
	},
});

export const { setAccessToken, clearAccessToken } = accessTokenSlice.actions;
export const selectAccessToken = (state: RootState) => state.accessToken;
export default accessTokenSlice.reducer;
