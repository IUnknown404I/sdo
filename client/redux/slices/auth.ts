import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';

export type SessionStateT = 'active' | 'expired' | 'closed';
export interface AuthReduxSliceI {
	state: boolean;
	sessionState: SessionStateT;
	lastPages?: string[];
	isBusy: boolean;
}

export const initialAuthState: AuthReduxSliceI = {
	state: false,
	sessionState: 'closed',
	lastPages: undefined,
	isBusy: false,
};

export const authSlice = createSlice({
	name: 'auth',
	initialState: initialAuthState,
	reducers: {
		*
	},
});

export const { setAuthState, setSessionState, setLastPages, setBusyState, clearAll } = authSlice.actions;
export const selectAuth = (state: RootState) => state.auth;
export const selectAuthBusy = (state: RootState) => state.auth.isBusy;
export default authSlice.reducer;
