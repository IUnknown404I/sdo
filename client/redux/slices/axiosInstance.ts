import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AxiosInstance } from 'axios';
import { RootState } from '../store';
import { generateAxiosInstance } from '../utils/axiosUtils';
import { setAccessToken } from './accessToken';

export const initialInstanceState = { instance: generateAxiosInstance({ setAccessToken: () => {}, clearBearer: () => {} }) };

export const axiosInstanceSlice = createSlice({
	name: 'axiosInstance',
	initialState: initialInstanceState,
	reducers: {
		*
	},
});

export const { setBearer, clearBearer } = axiosInstanceSlice.actions;
export const selectAxiosInstance = (state: RootState): AxiosInstance => state.axiosInstance?.instance;
export default axiosInstanceSlice.reducer;
