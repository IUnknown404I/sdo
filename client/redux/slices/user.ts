import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { UserFriendsI } from '../endpoints/chatEnd';
import { RootState } from '../store';

export const initialUserState: UserSliceI = {
	_systemRole: 'user',
	_permittedSystemRoles: ['user'],
};

export const userSlice = createSlice({
	name: 'user',
	initialState: initialUserState,
	reducers: {
		setUserAttributes: (state, action: PayloadAction<UserSliceI>) => {
			for (const attr in action.payload) {
				(state[attr as keyof UserSliceI] as unknown) = action.payload[attr as keyof UserSliceI];
			}
		},
		setUserDataFromPayload: (state, action: PayloadAction<AccessTokenPayload>) => {
			if (Object.keys(action.payload).length === 0) return;
			for (const attr in action.payload) {
				if (attr === '_systemRole' && isSystemRoleOption(action.payload._systemRole))
					state._systemRole = action.payload._systemRole;
				else
					(state[attr as keyof AccessTokenPayload] as unknown) =
						action.payload[attr as keyof AccessTokenPayload];
			}
		},
		personal: (state, action: PayloadAction<UserPersonalT>) => {
			if (Object.keys(action.payload).length === 0) return;
			for (const attr in action.payload) {
				if (state.personal === undefined) state.personal = {};
				state.personal[attr as keyof UserPersonalT] = action.payload[attr as keyof UserPersonalT];
			}
		},
		clearUserData: state => {
			state = initialUserState;
		},
	},
});

export const { setUserDataFromPayload, clearUserData, setUserAttributes, personal } = userSlice.actions;
export const selectUser = (state: RootState): UserSliceI => state.user;
export default userSlice.reducer;

export interface AccessTokenPayload {
	_systemRole: keyof typeof SystemRolesOptions;
	_permittedSystemRoles: (keyof typeof SystemRolesOptions)[];
	username?: string;
	email?: string;
	createdAt?: string;
	lastModified?: string;
	lastLoginIn?: string;
	isBlocked?: boolean;
	blockReason?: string;
	isActive?: boolean;
	failedAttempts?: number;
	lastFingerprints?: string;
}

export interface UserSliceI {
	_systemRole: keyof typeof SystemRolesOptions;
	_permittedSystemRoles: (keyof typeof SystemRolesOptions)[];
	username?: string;
	email?: string;
	createdAt?: string;
	lastModified?: string;
	lastLoginIn?: string;
	isBlocked?: boolean;
	blockReason?: string;
	isActive?: boolean;
	failedAttempts?: number;
	personal?: UserPersonalT;
	metaInfo?: UserMetaInformationI;
	lastFingerprints?: string;

	// globalGroupsID: string[];
	// activeCoursesID: UserCoursesAttributeItemType[];
	// finishedCoursesID: UserCoursesAttributeItemType[];
}

export type UserPersonalT = {
	name?: string;
	surname?: string;
	city?: string;
	company?: string;
	position?: string;
	tel?: string;
	avatar?: string;
};

export type UserMetaInformationI = {
	theme: 'light' | 'dark';
	contactVisibility: boolean;
	pushStatus: boolean;
	prefferedCommunication: 'email' | 'tel' | 'service';
};

/**
 * could be 'system', 'self-registry' or
 * string as: 'createdby-${admin_username}' | 'importedby-${admin_username}'
 */
export type UserCreationType = string | 'system' | 'self-registry';

export const SystemRolesOptions = {
	superuser: { accessLevel: 6, translation: 'Владелец' },
	developer: { accessLevel: 5, translation: 'Разработчик' },
	admin: { accessLevel: 4, translation: 'Администратор' },
	tutor: { accessLevel: 3, translation: 'Куратор' },
	content: { accessLevel: 2, translation: 'Контент-мейкер' },
	teacher: { accessLevel: 1, translation: 'Преподаватель' },
	user: { accessLevel: 0, translation: 'Пользователь' },
};

export const isSystemRoleOption = (
	systemrole: string | keyof typeof SystemRolesOptions,
): systemrole is keyof typeof SystemRolesOptions => systemrole in SystemRolesOptions;

// only for user-admin layers use
// this type doesnt appear anythere else
export type UserFullInfoType = {
	*
};

export type UserLoyalityBlockType = {
	coins: number;
	energy: number;
	level: number;
	experience: number;
	acquiredAwards: Array<{ awid: string; timestamp: number }>;
};
