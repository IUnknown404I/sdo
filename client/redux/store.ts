import { combineReducers, configureStore, ThunkMiddleware } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { useDispatch } from 'react-redux';
import { reducer as formReducer } from 'redux-form';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import thunk from 'redux-thunk';
import { checkProductionMode } from '../utils/utilityFunctions';
import { rtkApi } from './api';
import rootSaga from './sagas/rootSaga';
import { accessTokenSlice } from './slices/accessToken';
import { authSlice } from './slices/auth';
import { axiosInstanceSlice } from './slices/axiosInstance';
import { messengerSlice } from './slices/messenger';
import { userSlice } from './slices/user';
import { coursesSlice } from './slices/courses';

export const rootEndpoint = [rtkApi];

const rootReducer = combineReducers({
	form: formReducer,
	// redux
	auth: authSlice.reducer,
	user: userSlice.reducer,
	accessToken: accessTokenSlice.reducer,
	axiosInstance: axiosInstanceSlice?.reducer,
	messenger: messengerSlice.reducer,
	courses: coursesSlice.reducer,
	// api
	[rtkApi.reducerPath]: rtkApi.reducer,
	// ...rootEndpoint.reduce((accum, endpoint) => {
	// 	accum[endpoint.reducerPath as string] = endpoint.reducer;
	// 	return accum;
	// }, {} as { [key: string]: typeof rtkApi.reducer }),
});

export type ExtraArgumentsOptions = { productionMode: boolean };
const extraArgument: ExtraArgumentsOptions = {
	productionMode: checkProductionMode(),
};

const sagaMiddleware = createSagaMiddleware();
export const store = configureStore({
	reducer: rootReducer,
	middleware: getDefaultMiddleware =>
		getDefaultMiddleware({
			thunk: { extraArgument },
			serializableCheck: {
				// Ignore these action types
				// ignoredActions: ['auth/fetchAndUpdateTokens'],
				// Ignore these field paths in all actions
				// ignoredActionPaths: ['auth/fetchAndUpdateTokens'],
				// Ignore these paths in the state
				ignoredPaths: ['axiosInstance.instance'],
			},
		})
			.prepend()
			.concat(thunk)
			.concat(sagaMiddleware)
			.concat(rtkApi.middleware)
			// .concat(rootEndpoint.map(endpoint => endpoint.middleware as ThunkMiddleware))
			.concat(extraArgument.productionMode ? [] : [logger as ThunkMiddleware]),
});
// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
setupListeners(store.dispatch);
sagaMiddleware.run(rootSaga);

// Infer the `RootState` and `AppDispatch` types from the root Reducer itself
export type RootState = ReturnType<typeof rootReducer>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootStoreState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
