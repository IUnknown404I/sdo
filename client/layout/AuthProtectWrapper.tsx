import Box from '@mui/material/Box';
import { useTheme } from '@mui/material/styles';
import { unwrapResult } from '@reduxjs/toolkit';
import { useRouter } from 'next/router';
import React, { ReactNode } from 'react';
import { notification } from '../components/utils/notifications/Notification';
import useLoading from '../hooks/useLoading';
import { LoginFullObjI, rtkApi } from '../redux/api';
import { useTypedDispatch, useTypedSelector } from '../redux/hooks';
import { clearAccessToken } from '../redux/slices/accessToken';
import { setAuthState } from '../redux/slices/auth';
import { clearBearer } from '../redux/slices/axiosInstance';
import { clearUserData } from '../redux/slices/user';
import AuthThunks from '../redux/thunks/auth';
import { checkTokenValidityWithDate } from '../utils/http';
import logapp from '../utils/logapp';

const ALLOWED_PATHS: string[] = [
	'login',
	'registration',
	'recovery',
	'replacement',
	'refresh-token',
	'password-rewrite',
	'error',
	'404',
];

/**@deprecated */
const allowedURIPaths = {
	upper: ['login', 'registration', 'recovery', 'refresh-token', 'password-rewrite', '404'],
	inner: ['error', 'recovery'],
};

/**
 * @IUnknown404I AuthProtectWrapper with authenticate, authorize, auto-refreshes and e.t.c. logic inside.
 * @param {JSX.Element | JSX.Element[]} children as inherited nodes.
 * @returns a protected wrapper with all auth logic for use across the application.
 */
const AuthProtectWrapper = (props: { children: ReactNode | ReactNode[] }): JSX.Element => {
	const theme = useTheme();
	const router = useRouter();
	const dispatch = useTypedDispatch();

	const auth = useTypedSelector(state => state.auth.state);
	const userDTO = useTypedSelector(state => state.user);
	const accessTokenDTO = useTypedSelector(state => state.accessToken);

	const isFetching = React.useRef<boolean>(false);
	const [cookiesClearReq, _mutationData] = rtkApi.useClearCookiesMutation();

	const { Loader } = useLoading({
		size: 50,
		value: 25,
		disableShrink: false,
		animationDuration: '2.5s',
	});
	const { Loader: SecondLoader } = useLoading({
		size: 64,
		value: 100,
		disableShrink: false,
		animationDuration: '2.5s',
	});
	const { Loader: ThirdLoader } = useLoading({
		size: 50,
		value: 25,
		disableShrink: false,
		animationDuration: '2.5s',
	});

	React.useEffect(() => {
		// clear the previous request data if necessary
		checkPreviousRequestedData();

		// stop if auth-check logic is already running
		if (isFetching.current) return;
		isFetching.current = true;

		new Promise<void>((resolve, reject) => {
			// force redirect from public routes to inner app if user already logged-in
			if (auth && (router.route === '/login' || router.route === '/registry' || router.route === '/recovery')) {
				if (checkAuthValidity()) {
					const requestedPath = localStorage.getItem('initial-requested-path');
					// if user tried to go to the auth-secured page, log him in and redirect to the page
					if (!!requestedPath) {
						try {
							checkPreviousRequestedData(true);
							router.push(requestedPath.split(' ')[0]);
							return;
						} catch (error) {}
					}
					// either only redirect to the main inner page
					redirectToApp();
					resolve();
				}
			}

			const isPublicRoute = checkUrlPath(router.asPath);
			// public route and no auth --> revalidating
			if (isPublicRoute && (!auth || !checkAuthValidity())) refreshLogic();
			// not auth and private route requested --> refreshing
			else if (!auth && !isPublicRoute) refreshLogic();
			// auth and private route --> revalidating
			// else if (auth && !isPublicRoute && !checkAuthValidity()) redirectToLoginPage();
			else if (auth && !isPublicRoute && !checkAuthValidity()) refreshLogic();
			resolve();
		});

		isFetching.current = false;
		return () => {
			checkPreviousRequestedData();
		};
	}, [auth, router.asPath]);

	if (auth || checkUrlPath(router.asPath)) return <>{props.children}</>;
	else
		return (
			<Box
				sx={{
					position: 'absolute',
					inset: '0 0',
					display: 'flex',
					flexDirection: 'column',
					alignItems: 'center',
					justifyContent: 'center',
					gap: '1.25rem',
					backgroundColor: theme => (theme.palette.mode === 'light' ? '#ffffff' : '#162433'),
				}}
			>
				<Box
					sx={{
						position: 'relative',
						display: 'flex',
						flexDirection: 'row',
						alignItems: 'center',
						gap: '1.5rem',
					}}
				>
					{Loader} {SecondLoader} {ThirdLoader}
				</Box>
			</Box>
		);

	/**
	 * @IUnknown404I Check the validity of auth state by listing userDTO and validating access-token attributes.
	 * @description doesn't logout the user!
	 * @returns boolean flag of user's auth validity.
	 */
	function checkAuthValidity(): boolean {
		try {
			if (
				!userDTO ||
				!(!!userDTO.username && !!userDTO.email) ||
				!accessTokenDTO ||
				!accessTokenDTO.access_token ||
				!checkTokenValidityWithDate({
					access_token: accessTokenDTO.access_token,
					expires_in: accessTokenDTO.expires_in,
				})
			) {
				logapp.log('[error] checkAuthValidity FAILED');
				// refreshLogic();
				return false;
			} else {
				logapp.log('[ok] checkAuthValidity is OK');
				return true;
			}
		} catch (e) {
			return false;
		}
	}

	/**
	 * @IUnknown404I Refresh auth, session and tokens states on fullfilled state duo session state to "expired" and redirecting to the login page.
	 */
	function refreshLogic(): void {
		dispatch(AuthThunks.fetchAndUpdateTokens())
			.then(unwrapResult)
			.then(data => {
				if (
					(typeof data !== 'object' && typeof data !== 'string') ||
					(typeof data === 'object' && 'name' in data && data.name === 'AxiosError')
				) {
					logapp.info('AUTH WRAPPER -> reauth thunk throws an error, result = ', data);
					redirectToLoginPage();
					return;
				} else if (typeof data === 'string') {
					logapp.info('AUTH WRAPPER -> reauth thunk stopped cause the Auth-Busy-State', data);
					return;
				}

				logapp.info('AUTH WRAPPER -> reauth thunk works good, result = ', data);
				logapp.info(data as LoginFullObjI);
			})
			.catch(() => {
				logapp.info('AUTH WRAPPER -> reauth thunk returned an error!');
				redirectToLoginPage();
			});
	}

	/**
	 * @IUnknown404I
	 * Redirecting to the login page with signing out and clearing tokens if needed. Also the auth-warn toast will occure.
	 */
	function redirectToLoginPage(): void {
		if (checkUrlPath(router.asPath)) return;

		// saving the requested path for next authorization in this session
		if (
			router.asPath !== '/' &&
			router.asPath !== '/login' &&
			router.asPath !== '/registry' &&
			router.asPath !== '/recovery'
		)
			localStorage.setItem('initial-requested-path', router.asPath + ' ' + Date.now());

		cookiesClearReq.call('', '');
		dispatch(clearUserData());
		dispatch(clearBearer());
		dispatch(clearAccessToken());
		if (auth) dispatch(setAuthState(false));

		router.push('/login').then(async () =>
			setTimeout(() => {
				const message = 'Вы не авторизованы! Войдите в систему для доступа к запрошенному ресурсу.';
				const displayedToasts = document?.querySelectorAll('.Toastify__toast');
				let isDublicate = false;

				if (displayedToasts.length)
					displayedToasts.forEach(toast => (toast.innerHTML.includes(message) ? (isDublicate = true) : {}));

				logapp.log('[%] found Toast dublicate. Aborting notification.');
				if (!isDublicate)
					notification({
						message,
						type: 'error',
						autoClose: 6000,
						theme: theme.palette.mode === 'dark' ? 'dark' : 'light',
					});
			}, 1000),
		);
	}

	/**
	 * @IUnknown404I If provided URI is in public zone, will return "true", for inner views - "false".
	 * @param uri current page's full urlPath or in '/[path_after_domain]' syntaxis.
	 * @returns boolean flag is in public zone.
	 */
	function checkUrlPath(uri: string): boolean {
		if (uri === undefined) return false;
		const splittedURI = (uri.startsWith('/') ? uri.slice(1) : uri)
			.split('/')
			.map(el => (el.includes('?') ? el.slice(0, el.indexOf('?')) : el));
		return splittedURI.reduce((flag, cur) => flag && ALLOWED_PATHS.includes(cur), true);
	}

	/**
	 *  redirect to the inner main page
	 */
	function redirectToApp(): void {
		router.push('/').then();
	}

	/**
	 * Checks the prpevious requested data by user and clears it if 10 minutes passed since the request.
	 * @param clear boolean flag (false by default) for forced clean the data.
	 * @returns void
	 */
	function checkPreviousRequestedData(clear: boolean = false): void {
		const previousRequest = localStorage.getItem('initial-requested-path');
		if (!previousRequest) return;

		const previousTime = previousRequest.split(' ')[1];
		if (clear || Date.now() - parseInt(previousTime) > 10 * 60 * 1000)
			localStorage.setItem('initial-requested-path', '');
	}
};

export default AuthProtectWrapper;
