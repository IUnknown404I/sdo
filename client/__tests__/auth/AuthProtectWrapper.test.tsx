/**
 * @jest-environment jsdom
 */

import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import AuthProtectWrapper from '../../layout/AuthProtectWrapper';
import MetaWrapper from '../../layout/MetaWrapper';
import { initialAccessTokenState } from '../../redux/slices/accessToken';
import { initialAuthState } from '../../redux/slices/auth';
import { initialInstanceState } from '../../redux/slices/axiosInstance';
import { initialUserState } from '../../redux/slices/user';
import { store } from '../../redux/store';
import ToggleColorMode from '../../theme/Theme';

jest.mock('next/router', () => ({
	useRouter: jest.fn(() => ({
		query: {},
		route: '/login',
		asPath: 'login',
		push: () => new Promise(() => true),
	})),
}));
const mockStore = configureMockStore([thunk]);

const expectedElement = (
	<span
		aria-label=''
		className='MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-185k7aw-MuiCircularProgress-root'
		role='progressbar'
		style={{ width: '64px', height: '64px' }}
	>
		<svg className='MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg' viewBox='22 22 44 44'>
			<circle
				className='MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle'
				cx='44'
				cy='44'
				fill='none'
				r='20'
				stroke-width='4'
			/>
		</svg>
	</span>
);

beforeEach(() => {
	jest.mock('next/router', () => ({
		useRouter: jest.fn(() => ({
			query: {},
			asPath: 'login',
		})),
	}));
});

describe('AuthProtectWrapper ->', () => {
	it('correct renders AuthProtectWrapper:', () => {
		const { container, getByText } = render(
			<Provider store={store}>
				<ToggleColorMode>
					<MetaWrapper>
						<AuthProtectWrapper>
							<ToastContainer
								position='top-right'
								autoClose={4500}
								hideProgressBar={false}
								newestOnTop={false}
								closeOnClick
								rtl={false}
								pauseOnFocusLoss
								draggable
								pauseOnHover
								theme='light'
							/>
							<p>Test render!</p>
						</AuthProtectWrapper>
					</MetaWrapper>
				</ToggleColorMode>
			</Provider>,
		);

		waitFor(() => expect(container).toMatchSnapshot());
		waitFor(() => expect(getByText('Test render!')).not.toBeInTheDocument());
		waitFor(() => expect(expectedElement as unknown as HTMLElement).toBeInTheDocument());
	});

	it('authorized state test:', () => {
		const mockedStore = mockStore({
			instance: initialInstanceState,
			auth: {
				...initialAuthState,
				state: true,
				sessionState: 'active',
			},
			accessToken: {
				...initialAccessTokenState,
				access_token:
					'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InJuYWRtaW4iLCJlbWFpbCI6IiIsImNvbXBhbnkiOiIiLCJpc0FjdGl2ZSI6dHJ1ZSwiaXNCbG9ja2VkIjpmYWxzZSwiYmxvY2tSZWFzb24iOiIiLCJsYXN0TG9naW5JbiI6IjAxLjAxLjIwMjMgMTU6NTAuMjgiLCJmYWlsZWRBdHRlbXB0cyI6MCwiY3JlYXRlZEF0IjoiMDEuMDEuMjAyMyAxNDo0Ny4zMyIsImlhdCI6MTY4NTUzOTUwOCwiZXhwIjoxOTg1NTM5Njg4fQ.UnjzG45xKdFWK6bbcd4ZKYeGPny0cD9CYbaA4HpFEhU',
				expires_in: '2033-05-31T13:30:43.978Z',
			},
			user: {
				...initialUserState,
				username: 'testone',
				email: 'testone@mail.ru',
				// privileges: ,
				company: 'test',
				createdAt: '01.01.20',
				lastModified: '01.01.23',
				lastLoginIn: '01.05.20',
				isBlocked: false,
				blockReason: '',
				isActive: true,
				failedAttempts: 0,
			},
		});

		const { container, getByText } = render(
			<Provider store={mockedStore}>
				<ToggleColorMode>
					<MetaWrapper>
						<AuthProtectWrapper>
							<ToastContainer
								position='top-right'
								autoClose={4500}
								hideProgressBar={false}
								newestOnTop={false}
								closeOnClick
								rtl={false}
								pauseOnFocusLoss
								draggable
								pauseOnHover
								theme='light'
							/>
							<p>Test render!</p>
						</AuthProtectWrapper>
					</MetaWrapper>
				</ToggleColorMode>
			</Provider>,
		);

		waitFor(() => expect(container).toMatchSnapshot());
		waitFor(() => expect(getByText('Test render!')).toBeInTheDocument());
	});
});
