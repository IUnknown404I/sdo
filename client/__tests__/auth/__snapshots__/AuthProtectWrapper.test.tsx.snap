// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProtectWrapper -> authorized state test: 1`] = `
<div>
  <div
    class="Toastify"
  />
  <p>
    Test render!
  </p>
</div>
`;

exports[`AuthProtectWrapper -> correct renders AuthProtectWrapper: 1`] = `
<div>
  <div
    class="Toastify"
  />
  <p>
    Test render!
  </p>
</div>
`;
